src/main/java/com/google/devtools/build/lib/remote/RemoteActionInputFetcher.java

```diff
@@ -173,23 +180,30 @@ void downloadFile(Path path, FileArtifactValue metadata)
         Context prevCtx = ctx.attach();
         try {
           Digest digest = DigestUtil.buildDigest(metadata.getDigest(), metadata.getSize());
-          download = remoteCache.downloadFile(path, digest);
+          Path tmpPath = toTmpDownloadPath(path);
+          download = remoteCache.downloadFile(tmpPath, digest);
           downloadsInProgress.put(path, download);
           Futures.addCallback(
               download,
               new FutureCallback<Void>() {
                 @Override
                 public void onSuccess(Void v) {
-                  synchronized (lock) {
-                    downloadsInProgress.remove(path);
-                    downloadedPaths.add(path);
-                  }
-
                   try {
-                    path.chmod(0755);
+                    tmpPath.chmod(0755);
                   } catch (IOException e) {
                     logger.atWarning().withCause(e).log("Failed to chmod 755 on %s", path);
                   }
+                  try {
+                    FileSystemUtils.moveFile(tmpPath, path);
+                  } catch (IOException e) {
+                    logger.atWarning().withCause(e).log(
+                      "Failed to move to final location %s. -> %s", tmpPath, path
+                    );
+                  }
+                  synchronized (lock) {
```

Comment:
Moved the synchronize block below. A file is considered to be downloaded only after its permissions are set and it has been moved to its final location.

Reply:
I think this is wrong. You are doing this even when the above code fails with an IOException